drv/vsc7448/src/port.rs

```diff
@@ -0,0 +1,166 @@
+
+    port_flush_inner(port.into(), v)?;
+
+    // 10: Reset the MAC clock domain
+    v.modify(dev10g.DEV_CFG_STATUS().DEV_RST_CTRL(), |r| {
+        r.set_pcs_tx_rst(1);
+        r.set_mac_rx_rst(1);
+        r.set_mac_tx_rst(1);
+        r.set_speed_sel(6);
+    })?;
+
+    // 11: Clear flushing
+    v.modify(Vsc7448::HSCH().HSCH_MISC().FLUSH_CTRL(), |r| {
+        r.set_flush_ena(0);
+    })?;
+
+    // Bonus for 10G ports: disable XAUI, RXAUI, SFI PCS
+    v.modify(dev10g.PCS_XAUI_CONFIGURATION().PCS_XAUI_CFG(), |r| {
+        r.set_pcs_ena(0);
+    })?;
+    v.modify(dev10g.PCS2X6G_CONFIGURATION().PCS2X6G_CFG(), |r| {
+        r.set_pcs_ena(0);
+    })?;
+    v.modify(
+        Vsc7448::PCS10G_BR(dev.index()).PCS_10GBR_CFG().PCS_CFG(),
+        |r| {
+            r.set_pcs_ena(0);
+        },
+    )?;
+
+    Ok(())
+}
+
+/// Shared logic between 1G and 10G port flushing
+fn port_flush_inner(port: u32, v: &Vsc7448Spi) -> Result<(), VscError> {
+    // 3: Disable traffic being sent to or from switch port
+    v.modify(Vsc7448::QFWD().SYSTEM().SWITCH_PORT_MODE(port), |r| {
+        r.set_port_ena(0)
+    })?;
+
+    // 4: Disable dequeuing from the egress queues
+    v.modify(Vsc7448::HSCH().HSCH_MISC().PORT_MODE(port), |r| {
+        r.set_dequeue_dis(1)
+    })?;
+
+    // 5: Disable Flowcontrol
+    v.modify(Vsc7448::QSYS().PAUSE_CFG().PAUSE_CFG(port), |r| {
+        r.set_pause_ena(0)
+    })?;
+
+    // 5.1: Disable PFC
+    v.modify(Vsc7448::QRES().RES_QOS_ADV().PFC_CFG(port), |r| {
+        r.set_tx_pfc_ena(0)
+    })?;
+
+    // 6: Wait a worst case time 8ms (jumbo/10Mbit)
+    hl::sleep_for(8);
+
+    // 7: Flush the queues accociated with the port
+    v.modify(Vsc7448::HSCH().HSCH_MISC().FLUSH_CTRL(), |r| {
+        r.set_flush_port(port);
+        r.set_flush_dst(1);
+        r.set_flush_src(1);
+        r.set_flush_ena(1);
+    })?;
+
+    // 8: Enable dequeuing from the egress queues
+    v.modify(Vsc7448::HSCH().HSCH_MISC().PORT_MODE(port), |r| {
+        r.set_dequeue_dis(0)
+    })?;
+
+    // 9: Wait until flushing is complete
+    port_flush_wait(port, v)?;
+
+    Ok(())
+}
+
+/// Waits for a port flush to finish.  This is based on
+/// `jr2_port_flush_poll` in the MESA SDK
+fn port_flush_wait(port: u32, v: &Vsc7448Spi) -> Result<(), VscError> {
+    for _ in 0..32 {
+        let mut empty = true;
+        // DST-MEM and SRC-MEM
+        for base in [0, 2048] {
+            for prio in 0..8 {
+                let value = v.read(
+                    Vsc7448::QRES().RES_CTRL(base + 8 * port + prio).RES_STAT(),
+                )?;
+                empty &= value.maxuse() == 0;
+                // Keep looping, because these registers are clear-on-read,
+                // so it's more efficient to read them all, even if we know
+                // that the port isn't currently empty.
+            }
+        }
+        if empty {
+            return Ok(());
+        }
+        hl::sleep_for(1);
```

Comment:
Your other sleeps have comments justifying the duration choices. Do we have a justification for this one?

Do we know how often this has to retry? If the answer is >50%, should we sleep _before_ checking to reduce load?

Reply:
Bump this to 2000, which is what the SDK does, and leave a comment to that effect